### PR TITLE
Issue 1983: Expose latest connector version.

### LINKED ISSUE #1983: Separate Connector from Core release lifecycle
## Tell us about the problem you're trying to solve
Today users have very little visibility when new versions of a connector are released. When we release core, we do so with a fixed set of connector versions. This isn't really a problem. The issue is that if we release a new version of a connector after the fact, the user has no good way (in the UI) to figure out that there is a new connector available. 

They effectively get stuck on a connector version until they ask for help. We have numerous user issues due to connector version problems (e.g. [issue](https://github.com/airbytehq/airbyte/pull/2205)). This is a common occurrence because connectors evolve at a different pace than core. It is pretty common that a user will not need upgrade their current core version but will want to upgrade a connector version to get a bug fix or new feature.

## Describe the solution you’d like
I would like airbyte to be able to pull from a feed updates about connectors and detect:
- there is a new version and allow to update
- ~~ensure that core is compatible with the new version of the connector. This should be less of an issue as we stabilize the protocol.~~ (I propose we punt on this for now. It can be done on a follow on task with no additional overhead. The reason to punt is that it requires us to be more thoughtful about protocol version and we haven't figure that out yet, cgardens)

## Describe the alternative you’ve considered or used
Today we have an escape hatch where if a user knows that there is a new version of a connector, they can manually input its version number in the UI. This process requires them to either read our change log or find where we declare new versions of connectors in the code base. This process is not ideal.

Issue for the UI work: https://github.com/airbytehq/airbyte/issues/2218

### PR DESCRIPTION
**Pull request by davinchia**
## What
* Expose the latest connector version to users for better connector UX experience.
* The airbyte webapp currently displays the docker tag. Users can specify a custom tag to use within using the same cell. However, users do not know what the latest connector image tag is without manually checking the Airbyte github or change log.
* These changes will provide an endpoint for the webapp to retrieve the relevant information.

## How
- Add `listLatest` endpoint to the `SourceDefinition` and `DestinationDefinition` resources.
- Implement the relevant endpoints in the relevant handlers.
- Slight refactor of the `SeedRepository` class to minimise validation drift.
- Use of `MockWebServer` to test Java 11 Http Client.

For more detail see: https://docs.google.com/document/d/1CeqSw9ooWDZP0yvVSmi28DgIHrlW97Tlbm9lIDSzipQ/edit

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
* `config.yaml` (to understand api changes)
* `ConfigurationApi.java` (how handler is exposed to servelet)
* `DestinationDefinitionsHandler.java`, `SourceDefinitionHandler.java` (main changes)
* `SeedRepository.java` (minor refactoring)
* Test files.

### CODE REVIEW

**Comment by davinchia** [comment]
@cgardens currently only done with the `DestinationDefinition` bits. However, since the approach is similar for both, want your thoughts before I proceed with the `SourceDefinition` bits. Nothing too crazy - should be pretty standard.

I'm avoiding any sort of cross api interactions for now e.g. `listDestinationDefinition` refreshes itself by calling `listLatestDestinationDefinition` under the hood etc, since we haven't talked about that in detail yet. Also, think the caching stuff should happen in a follow up PR. What do you think?

Some minor stuff:
1) How do we feel about `var`? I like it, however understand if people are wary of introducing slightly different coding styles.
2) Same questions about `final`. I notice the files you touch tend to use `final` judiciously. What is our stance?
3) Understand people hate reviewing code without tests - mocking the final class is proving to be a struggle. Definitely will have tests before we merge in! Any tips?

Thanks!

File: airbyte-api/src/main/openapi/config.yaml
Comment (davinchia) [inline airbyte-api/src/main/openapi/config.yaml:201]: this seemed like a mistake to me. lmk if I misunderstood this.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/DestinationDefinitionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/DestinationDefinitionsHandler.java:154]: IIUC, a RTE automatically triggers a http 500 yes?

Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/DestinationDefinitionsHandler.java]: [This](https://github.com/airbytehq/airbyte/blob/master/airbyte-server/src/main/java/io/airbyte/server/errors/UncaughtExceptionMapper.java#L36) tells me it does.

**Comment by davinchia** [comment]
Built the new docker image, tested it locally and confirmed it works:

```
(⎈ |gke_causal-sky-277804_us-central1_autopilot-cluster-1:default)
➜  ~ curl -X POST "http://localhost:8001/api/v1/destination_definitions/list" -H  "accept: application/json" -d ""

{"destinationDefinitions":[{"destinationDefinitionId":"f7a7d195-377f-cf5b-70a5-be6b819019dc","name":"Redshift","dockerRepository":"airbyte/destination-redshift","dockerImageTag":"0.1.8","documentationUrl":"https://docs.airbyte.io/integrations/destinations/redshift"},{"destinationDefinitionId":"22f6c74f-5699-40ff-833c-4a879ea40133","name":"BigQuery","dockerRepository":"airbyte/destination-bigquery","dockerImageTag":"0.1.16","documentationUrl":"https://docs.airbyte.io/integrations/destinations/bigquery"},{"destinationDefinitionId":"af7c921e-5892-4ff2-b6c1-4a5ab258fb7e","name":"MeiliSearch","dockerRepository":"airbyte/destination-meilisearch","dockerImageTag":"0.1.0","documentationUrl":"https://docs.airbyte.io/integrations/destinations/meilisearch"},{"destinationDefinitionId":"a625d593-bba5-4a1c-a53d-2d246268a816","name":"Local JSON","dockerRepository":"airbyte/destination-local-json","dockerImageTag":"0.1.4","documentationUrl":"https://docs.airbyte.io/integrations/destinations/local-json"},{"destinationDefinitionId":"25c5221d-dce2-4163-ade9-739ef790f503","name":"Postgres","dockerRepository":"airbyte/destination-postgres","dockerImageTag":"0.1.16","documentationUrl":"https://docs.airbyte.io/integrations/destinations/postgres"},{"destinationDefinitionId":"8be1cf83-fde1-477f-a4ad-318d23c9f3c6","name":"Local CSV","dockerRepository":"airbyte/destination-csv","dockerImageTag":"0.1.8","documentationUrl":"https://docs.airbyte.io/integrations/destinations/local-csv"},{"destinationDefinitionId":"424892c4-daac-4491-b35d-c6688ba547ba","name":"Snowflake","dockerRepository":"airbyte/destination-snowflake","dockerImageTag":"0.1.18","documentationUrl":"https://docs.airbyte.io/integrations/destinations/snowflake"}]}%
(⎈ |gke_causal-sky-277804_us-central1_autopilot-cluster-1:default)
➜  ~ curl -X POST "http://localhost:8001/api/v1/destination_definitions/listLatest" -H  "accept: application/json" -d ""

{"destinationDefinitions":[{"destinationDefinitionId":"a625d593-bba5-4a1c-a53d-2d246268a816","name":"Local JSON","dockerRepository":"airbyte/destination-local-json","dockerImageTag":"0.1.4","documentationUrl":"https://docs.airbyte.io/integrations/destinations/local-json"},{"destinationDefinitionId":"8be1cf83-fde1-477f-a4ad-318d23c9f3c6","name":"Local CSV","dockerRepository":"airbyte/destination-csv","dockerImageTag":"0.1.8","documentationUrl":"https://docs.airbyte.io/integrations/destinations/local-csv"},{"destinationDefinitionId":"25c5221d-dce2-4163-ade9-739ef790f503","name":"Postgres","dockerRepository":"airbyte/destination-postgres","dockerImageTag":"0.1.16","documentationUrl":"https://docs.airbyte.io/integrations/destinations/postgres"},{"destinationDefinitionId":"22f6c74f-5699-40ff-833c-4a879ea40133","name":"BigQuery","dockerRepository":"airbyte/destination-bigquery","dockerImageTag":"0.1.16","documentationUrl":"https://docs.airbyte.io/integrations/destinations/bigquery"},{"destinationDefinitionId":"424892c4-daac-4491-b35d-c6688ba547ba","name":"Snowflake","dockerRepository":"airbyte/destination-snowflake","dockerImageTag":"0.1.18","documentationUrl":"https://docs.airbyte.io/integrations/destinations/snowflake"},{"destinationDefinitionId":"f7a7d195-377f-cf5b-70a5-be6b819019dc","name":"Redshift","dockerRepository":"airbyte/destination-redshift","dockerImageTag":"0.1.8","documentationUrl":"https://docs.airbyte.io/integrations/destinations/redshift"},{"destinationDefinitionId":"af7c921e-5892-4ff2-b6c1-4a5ab258fb7e","name":"MeiliSearch","dockerRepository":"airbyte/destination-meilisearch","dockerImageTag":"0.1.0","documentationUrl":"https://docs.airbyte.io/integrations/destinations/meilisearch"}]}%
```

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java:64]: Some of these stuff can be potentially pulled out into either a `JavaHttpClient` class, or a 'Github source of truth' model. Felt unnecessary given that most of our stuff I expect only two usages - SourceDefinition and DesintationDefinition in the short/medium term.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java:148]: Same comment here regarding duplication and potential consolidation into 'one' class.

File: airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java
Comment (cgardens) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java:80]: I'm not super convinced we need this. Would like to figure out why mocking wasn't working and see if we can fix that. If it's truly not possible then we can use this, but if mocking can work I'd prefer that.

e.g. this worked.
```
@Test
void testHttpClient() {
  final HttpResponse httpResponse = mock(HttpResponse.class);
  when(httpResponse.body()).thenReturn("hello");
  System.out.println("httpResponse.body() = " + httpResponse.body());
}
```

File: airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java
Comment (cgardens) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java:95]: didn't know this was a thing. cool!

**Review by cgardens — CHANGES_REQUESTED** [review]
Awesome first (real) PR!!! 🎉  

I'm requesting chances because there's 3 things I want to iron out: 
1. I think we can convert between the registry and the json objects all in one place (and test it once) instead of having it in 3 places. 
2. I don't think we should need this mock server in the tests. Want to make sure we know for sure we need it before adding it.
3. you left comments about it already, but i think it might be a good idea to encapsulate pulling the data from github in its own class for testability and for making sure that that logic is in one spot. This one can definitely go either way, but want to make sure we consider it before moving forward.

(i mention all 3 of these in more detail inline)

otherwise a bunch of smaller comments. let me know if anything doesn't make sense! Looking very good.

File: airbyte-api/src/main/openapi/config.yaml
Comment (davinchia) [reply airbyte-api/src/main/openapi/config.yaml:175]: That's fair. I like setting the expectation lower for now, and changing this when/if we decide the normal `list` route can cache stuff.

File: airbyte-api/src/main/openapi/config.yaml
Comment (davinchia) [reply airbyte-api/src/main/openapi/config.yaml:184]: Ah, didn't realise this!

File: airbyte-config/init/src/main/java/io/airbyte/config/init/SeedRepository.java
Comment (davinchia) [reply airbyte-config/init/src/main/java/io/airbyte/config/init/SeedRepository.java:86]: offline conclusion: we are good to go for now!

File: airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java
Comment (davinchia) [reply airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java:95]: thank you, thank you!

File: airbyte-config/init/src/main/java/io/airbyte/config/init/SeedRepository.java
Comment (davinchia) [reply airbyte-config/init/src/main/java/io/airbyte/config/init/SeedRepository.java:120]: My heuristic is the rule of three - and we have just hit it (Seed Repository, SourceDefinitionHandler and DestinationDefinitionHandler) so let's do it!

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SourceDefinitionsHandler.java:64]: I feel combining them now is slightly premature. IMO we don't get much yet since the functionality is nicely encapsulated in 2 - 3 functions within each file. I'd rather wait for when we have a 3rd use case to do it.

You have a better sense of the codebase + future use cases so I'll defer to you. If using Github as a data store is something we foresee and want to encourage then I feel creating a separate concept makes sense.

Ultimately I'm a 2/10 on this one so if you feel stronger I'm happy to do so!

File: airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java
Comment (davinchia) [reply airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java:80]: Let me spend half an hour poking around to see if I can replicate this.

Comment (davinchia) [reply airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java:80]: Some of my prev annoyance with mocking was from the previous async call and having to do future -> httpresponse -> body.

Yes, you are right. Managed to get the mock stuff to work. Not sure what I was running into previously..

I still have a slight preference for using `MockWebServer`. I quite like how mocking is minimal and the control we get, especially on headers and all.

This is what it'll look timeout test will look like with pure mockito. Thought it's concise, I don't like how the delay statement has so much in it haha.
```    var mockHttpClient = mock(HttpClient.class);
    var mockResponse = mock(HttpResponse.class);
    when(mockResponse.body()).thenReturn("whooo");
    
    doAnswer( new AnswersWithDelay( 1000,  new Returns(mockResponse)) ).when(mockHttpClient).send(any(), any());
    
    var request = HttpRequest.newBuilder().uri(URI.create("https://www.google.com")).build();
    var send = mockHttpClient.send(request, BodyHandlers.ofString());

    System.out.println(send.body());
    ```

Definitely personal opinion, happy to change everything to mock if you feel strongly.

**Comment by davinchia** [comment]
Thanks @cgardens!  

General thoughts with more details in comments:
1. Agree on centralising the conversion. Created the `RawListToStandardXDefinition` class.
2. Thanks for taking the time to play with the mock stuff! I must have been goofing on some syntax/local Intellij set up. I do think it's worth using `MockWebServer` as it simplifies some of the timeout stuff and has more concise control over header/body etc. That being said, I've never been a big fan of mocking so I def lean away from it more. Happy to change if you feel strongly.
3. Agree with needing to consider. IMO it's still early to do so unless we know other systems are also going to use github as a source of truth/want to encourage using Github as a source of truth.

Lmk what you think!

File: airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java
Comment (cgardens) [reply airbyte-server/src/test/java/io/airbyte/server/handlers/DestinationDefinitionsHandlerTest.java:80]: kk. let's try it as is and see how it goes. i think my current reservation is that i don't know what mockwebserver is doing under the hood to make this all work. it seems like we're introducing some magic into the stack. we can change it in the future if it poses problems.

File: airbyte-config/models/src/main/java/io/airbyte/config/helpers/RawListToStandardXDefinitions.java
Comment (cgardens) [inline airbyte-config/models/src/main/java/io/airbyte/config/helpers/RawListToStandardXDefinitions.java:121]: I think you can use `Jsons.object(...)` here.

File: airbyte-config/models/src/main/java/io/airbyte/config/helpers/RawListToStandardXDefinitions.java
Comment (cgardens) [inline airbyte-config/models/src/main/java/io/airbyte/config/helpers/RawListToStandardXDefinitions.java:56]: `YamlListToStandardDefinitions`? since it only accepts yaml? I know I put the X in my comment but I don't think it's necessary now that i'm looking at how you've done it. It is clear without it.

File: airbyte-config/models/src/test/java/io/airbyte/config/helpers/RawListToStandardXDefinitionsTest.java
Comment (cgardens) [inline airbyte-config/models/src/test/java/io/airbyte/config/helpers/RawListToStandardXDefinitionsTest.java:93]: by convention methods for tests should contain the word test as a prefix or suffix.

File: airbyte-config/init/src/main/java/io/airbyte/config/init/SeedRepository.java
Comment (cgardens) [reply airbyte-config/init/src/main/java/io/airbyte/config/init/SeedRepository.java]: We end up at the same conclusion here, but just wanted to share a thought. I use this heuristic you mentioned alot too. I would actually say it's maybe not a good fit for this case though. 

Essentially we have are converting from a poorly defined object to a well defined object and coming up with a protocol on how to do it. The key in my mind that the because there's a poorly defined thing on one side (some raw yaml that we don't have an official struct for), it's easy to make a mistake if the code is duplicated because the compiler can't really help us. So my intuition here is that you don't want to centralize this logic and not have any duplication, because it is logic that is easy to mess up. and if the duplicated code does go out of sync, it will be hard to detect.

**Review by cgardens — APPROVED** [review]
Changes look good. One the 3 main points, I think 2 are resolved. For the one about combining fetching the records from github, I would like you to try out what combining it looks like. I think it's actually going to be pretty valuable. Let more info below on why. Anyway, I'd say try it, and if it's not better in your opinion stick with what you've got.

Otherwise once you do that and get the building passing, I think this is good to go! Nice work.

File: airbyte-config/models/src/main/java/io/airbyte/config/helpers/RawListToStandardXDefinitions.java
Comment (davinchia) [reply airbyte-config/models/src/main/java/io/airbyte/config/helpers/RawListToStandardXDefinitions.java:121]: Nice catch!

**Pull request merged**
